src/kakarot/instructions/stop_and_arithmetic_operations.cairo

```diff
@@ -63,468 +44,374 @@ namespace StopAndArithmeticOperations {
-    // @return ExecutionContext The pointer to the execution context.
     func exec_signextend{
         syscall_ptr: felt*,
         pedersen_ptr: HashBuiltin*,
         range_check_ptr,
         bitwise_ptr: BitwiseBuiltin*,
     }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
         alloc_locals;
+        let opcode = Internals.get_opcode(0x0b);
+        return Internals.exec_arithmetic_operation(ctx, opcode);
+    }
+}
+
+namespace Internals {
+    func get_opcode(index: felt) -> model.Opcode* {
+        let (opcode: model.Opcode*) = get_label_location(opcodes);
+        return opcode + index * model.Opcode.SIZE;
+
+        // See model.Opcode
+        // number
+        // gas
+        // stack_input
+        // stack_diff
+        // static_disabled
+        opcodes:
+        // STOP;
+        dw 0x00;
+        dw 0;
+        dw 0;
+        dw 0;
+        dw 0;
+        // ADD;
+        dw 0x01;
+        dw 3;
+        dw 2;
+        dw -1;
+        dw 0;
+        // MUL;
+        dw 0x02;
+        dw 5;
+        dw 2;
+        dw -1;
+        dw 0;
+        // SUB;
+        dw 0x03;
+        dw 3;
+        dw 2;
+        dw -1;
+        dw 0;
+        // DIV;
+        dw 0x04;
+        dw 5;
+        dw 2;
+        dw -1;
+        dw 0;
+        // SDIV;
+        dw 0x05;
+        dw 5;
+        dw 2;
+        dw -1;
+        dw 0;
+        // MOD;
+        dw 0x06;
+        dw 5;
+        dw 2;
+        dw -1;
+        dw 0;
+        // SMOD;
+        dw 0x07;
+        dw 5;
+        dw 2;
+        dw -1;
+        dw 0;
+        // ADDMOD;
+        dw 0x08;
+        dw 8;
+        dw 3;
+        dw -2;
+        dw 0;
+        // MULMOD;
+        dw 0x09;
+        dw 8;
+        dw 3;
+        dw -2;
+        dw 0;
+        // EXP;
+        dw 0x0A;
+        dw 10;
+        dw 2;
+        dw -1;
+        dw 0;
+        // SIGNEXTEND;
+        dw 0x0B;
+        dw 5;
+        dw 2;
+        dw -1;
+        dw 0;
+    }
 
-        let stack_underflow = is_le(ctx.stack.size, 1);
+    func exec_arithmetic_operation{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*, opcode: model.Opcode*) -> model.ExecutionContext* {
+        alloc_locals;
+        local stack: model.Stack*;
+
+        let stack_underflow = is_le(ctx.stack.size, opcode.stack_input - 1);
         if (stack_underflow != 0) {
             let (revert_reason_len, revert_reason) = Errors.stackUnderflow();
             let ctx = ExecutionContext.stop(ctx, revert_reason_len, revert_reason, TRUE);
             return ctx;
         }
 
-        // Stack input:
-        // 0 - a: number.
-        // 1 - b: exponent.
-        let stack = ctx.stack;
-        let (stack, popped) = Stack.pop_n(self=stack, n=2);
+        let (_stack, popped) = Stack.pop_n(ctx.stack, opcode.stack_input);
+
+        assert stack = _stack;
```

Comment:
can we add comments for all the dark cairo magic we do for readability and maintanability